Eliminate LoadOps.FROMCPU

kposborne2 opened the pull request:
- new `LazyBuffer.fromCPU` method to create from `ndarray` + changing the `__init__` method to have optional `LazyOp`/`RawBuffer`
- Add `int32` support for `RawMallocBuffer`

tinyb0t commented:
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/lazy.py                          20              17
tinygrad/ops.py                            1               1
tinygrad/runtime/lib.py                    7               1
tinygrad/tensor.py                         1               2
------------------------------------------------------------
total                                     29              21
------------------------------------------------------------
lines added in the tinygrad folder: 8
```

geohot commented:
This is +8 lines. It should be -10 lines.

kposborne2 commented:
Were you thinking of a different overall approach? Or is there other code that can be also removed now that this capability is available? I didn't really see one that could actually reduce the line count by 10 - if I remove the bugfix for empty tensors, inline the LazyBuffer construction method to tensor `__init__` (requires `ShapeTracker` import), and remove the assertion, it is still only -2 on the net, didn't really see a way to get to -10

geohot commented on tinygrad/tensor.py:
```diff
@@ -44,8 +45,7 @@ def __init__(self, data:Union[int, float, list, LazyBuffer, np.ndarray], device=
     elif isinstance(data, np.ndarray):
-      # TODO: create CPUBuffer directly
-      lazydata = LazyBuffer.loadop(LoadOps.FROMCPU, data.shape, dtypes.from_np(data.dtype), device, data)
+      lazydata = LazyBuffer(device, ShapeTracker(tuple(data.shape)), LoadOps, Device[device].buffer.fromCPU(data, **Device.extra_args(device)), dtypes.from_np(data.dtype))
```
Create a CPUBuffer then you can copy it later.

kposborne2 replied:
I see, so it should still defer the load if the device is not the cpu, that can simplify it

geohot commented on tinygrad/lazy.py:
```diff
@@ -87,32 +87,29 @@ def create_lazybuffer(device:str, shape:Union[ShapeTracker, Tuple[int, ...]], op
 class LazyBuffer:
   __deletable__ = ('op',)
-  def __init__(self, device:str, st:ShapeTracker, optype:OpType, op:LazyOp, dtype:DType):
+  def __init__(self, device:str, st:ShapeTracker, optype:OpType, src:Union[LazyOp, RawBuffer], dtype:DType):
```
I'd make a different constructor for this.

kposborne2 replied:
WDYT of a `fromCPU` method that takes the `np.ndarray`? The current `RawBuffer` does not remember what was its device

geohot commented:
Nice. What about that TODO?

kposborne2 commented:
> Nice. What about that TODO?

In the original implementation, I added a new EmptyBuffer class to handle buffers with 0 size, I reverted that to reduce the lines added. It is contained in this diff here: https://github.com/geohot/tinygrad/pull/920/commits/bbdcdc8713f60725012d9f1602e18de4bdd3a4ed, does that solution make sense to you? Note that this type of tensor is already broken in master if you try to call `.realize()` on it.

geohot commented:
Hmm, I don't think a buffer should ever be 0 size, just 0 dim.

kposborne2 commented:
I see, since the loading for non-cpu devices is deferred again, it won't hit the issue, so I will restore that test

geohot commented:
Nice work!

geohot commented:
Hmm, this broke docs/abstractions.py (which isn't tested in CI)

Want to fix and add to tests?

kposborne2 commented:
> Hmm, this broke docs/abstractions.py (which isn't tested in CI)
> 
> Want to fix and add to tests?

Yes, I can do that

geohot commented:
Done!

narangkay commented:
Note that without a bump to the pip release, running (now updated) abstractions.py in colab gives me

```
LazyOp(op=<LoadOps.FROMCPU: 1>, src=(), arg=<tinygrad.helpers.LazyNumpyArray object at 0x7fd0e8112140>)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-7-7442bbed48f4>](https://localhost:8080/#) in <cell line: 3>()
      1 print(lazyop.src[0].op)
      2 assert lazyop.src[0].op.op == LoadOps.FROMCPU
----> 3 assert lazyop.src[0].op.src[0].realized.toCPU()[0] == 2, "the arg of the FROM LazyOP is a LazyBuffer holding [2.]"
      4 assert result.lazydata.realized is None, "the LazyBuffer is not realized yet"

IndexError: tuple index out of range
``` 

Might be worth adding some sort of version check in abstractions.py
